plugins/GSdx/GSFastList.h

```diff
@@ -0,0 +1,253 @@
+		}
+
+		// Pop a free index from the stack
+		const uint16 free_index = m_free_indexes_stack[m_free_indexes_stack_top++];
+		m_buffer[free_index].data = data;
+		ListInsertFront(free_index);
+		return free_index;
+	}
+
+	__forceinline void push_front(const T& data) {
+		InsertFront(data);
+	}
+
+	__forceinline const T& back() const {
+		return m_buffer[LastIndex()].data;
+	}
+
+	__forceinline void pop_back() {
+		EraseIndex(LastIndex());
+	}
+
+	__forceinline uint16 size() const {
+		return m_free_indexes_stack_top;
+	}
+
+	__forceinline bool empty() const {
+		return size() == 0;
+	}	
+
+	__forceinline void EraseIndex(const uint16 index) {
+		ListRemove(index);
+		m_free_indexes_stack[--m_free_indexes_stack_top] = index;
+	}
+
+	__forceinline void MoveFront(const uint16 index) {
+		if (FirstIndex() != index) {
+			ListRemove(index);
+			ListInsertFront(index);
+		}
+	}
+
+	__forceinline const FastListIterator<T> begin() const {
+		return FastListIterator<T>(this, FirstIndex());
+	}
+
+	__forceinline const FastListIterator<T> end() const {
+		return FastListIterator<T>(this, 0);
+	}
+
+	__forceinline FastListIterator<T> erase(FastListIterator<T> i) {
+		EraseIndex(i.Index());
+		return ++i;
+	}
+
+private:
+	// Accessed by FastListIterator<T> using class friendship
+	__forceinline const T& Data(const uint16 index) const {
+		return m_buffer[index].data;
+	}
+
+	// Accessed by FastListIterator<T> using class friendship
+	__forceinline uint16 NextIndex(const uint16 index) const {
+		return m_buffer[index].next_index;
+	}
+
+	// Accessed by FastListIterator<T> using class friendship
+	__forceinline uint16 PrevIndex(const uint16 index) const {
+		return m_buffer[index].prev_index;
+	}
+
+	__forceinline uint16 FirstIndex() const {
+		return m_buffer[0].next_index;
+	}
+
+	__forceinline uint16 LastIndex() const {
+		return m_buffer[0].prev_index;
+	}
+
+	__forceinline bool Full() const {
+		// The minus one is due to the presence of the auxiliary element
+		return size() == m_capacity - 1;
+	}
+
+	__forceinline void ListInsertFront(const uint16 index) {
+		// Update prev / next indexes to add m_buffer[index] to the chain
+		Element<T>& head = m_buffer[0];
+		m_buffer[index].prev_index = 0;
+		m_buffer[index].next_index = head.next_index;
+		m_buffer[head.next_index].prev_index = index;
+		head.next_index = index;
+	}
+
+	__forceinline void ListRemove(const uint16 index) {
+		// Update prev / next indexes to remove m_buffer[index] from the chain
+		const Element<T>& to_remove = m_buffer[index];
+		m_buffer[to_remove.prev_index].next_index = to_remove.next_index;
+		m_buffer[to_remove.next_index].prev_index = to_remove.prev_index;
+	}
+
+	void Grow() {
+		if (m_capacity == USHRT_MAX) {
+			throw std::runtime_error("FastList size maxed out at USHRT_MAX (65535) elements, cannot grow futhermore.");
+		}
+
+		// 32768 is the first number that when doubled overflows uint16
+		const uint16 new_capacity = m_capacity < 32768 ? m_capacity * 2 : USHRT_MAX;
```

Comment:
I would then say: `m_capacity <= (uint16)(USHRT_MAX / 2)`, as `32768` is `(USHRT_MAX / 2) + 1`. Do you think the cast is necessary? (not that is makes any difference)